src/vdaf/prio3.rs

```diff
@@ -252,14 +253,131 @@ pub struct Prio3InputShare<F> {
+            (Share::Leader(input_share_data), Share::Leader(proof_share_data)) => {
+                for x in input_share_data {
+                    bytes.extend((*x).into());
+                }
+                for x in proof_share_data {
+                    bytes.extend((*x).into());
+                }
+            }
+            (Share::Helper(input_share_seed), Share::Helper(proof_share_seed)) => {
+                bytes.extend_from_slice(input_share_seed.as_slice());
+                bytes.extend_from_slice(proof_share_seed.as_slice());
+            }
+            _ => panic!("tried to encode input share with ambiguous encoding"),
+        };
+        bytes.extend_from_slice(self.joint_rand_seed_hint.as_slice());
+        bytes.extend_from_slice(self.blind.as_slice());
+    }
+
+    pub(crate) fn read_leader<R, A, T>(vdaf: &Prio3<T, A>, reader: &mut R) -> Option<Self>
+    where
+        R: Read,
+        A: Debug,
+        T: Type<Field = F>,
+    {
+        Self::read(true, vdaf, reader)
+    }
+
+    pub(crate) fn read_helper<R, A, T>(vdaf: &Prio3<T, A>, reader: &mut R) -> Option<Self>
+    where
+        R: Read,
+        A: Debug,
+        T: Type<Field = F>,
+    {
+        Self::read(false, vdaf, reader)
+    }
+
+    fn read<R, A, T>(is_leader: bool, vdaf: &Prio3<T, A>, reader: &mut R) -> Option<Self>
+    where
+        R: Read,
+        A: Debug,
+        T: Type<Field = F>,
+    {
+        let (input_share, proof_share) = match is_leader {
+            true => (
+                Share::read_leader(vdaf.typ.input_len(), reader)?,
+                Share::read_leader(vdaf.typ.proof_len(), reader)?,
+            ),
+            _ => (
+                Share::read_helper(vdaf.suite, reader)?,
+                Share::read_helper(vdaf.suite, reader)?,
+            ),
+        };
+
+        let mut joint_rand_seed_hint = Key::uninitialized(vdaf.suite);
+        let mut blind = Key::uninitialized(vdaf.suite);
+        reader
+            .read_exact(joint_rand_seed_hint.as_mut_slice())
+            .ok()?;
+        reader.read_exact(blind.as_mut_slice()).ok()?;
+
+        Some(Prio3InputShare {
+            input_share,
+            proof_share,
+            joint_rand_seed_hint,
+            blind,
+        })
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+/// The verification message emitted by each aggregator during the Prepare process.
+pub struct Prio3PrepareMessage<F> {
+    /// (A share of) the FLP verifier message. (See [`Type`](crate::pcp::Type).)
+    pub verifier: Vec<F>,
+
+    /// (A share of) the joint randomness seed.
+    //
+    // TODO(cjpatton) If `input.joint_rand_len() == 0`, then we don't need to bother with the joint
+    // randomness seed at all. See https://github.com/cjpatton/vdaf/issues/15.
+    pub joint_rand_seed: Key,
+}
+
+#[allow(dead_code)]
+impl<F: FieldElement> Prio3PrepareMessage<F> {
+    pub(crate) fn append(&self, bytes: &mut Vec<u8>) {
```

Comment:
As in the `Prio3InputShare` case, I think this should be `append<W: Write>(&self, &mut W) -> Result<(), VdafError> {}`